File: origin/YuGiOhCard/View/CardAttributePicker.m
#import "CardAttributePicker.h"
#import "StringConsts.h"

@interface CardAttributePicker() {
    UILabel * toolTitle;
    UIButton * toolButton;
}
@end

@implementation CardAttributePicker

- (id)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        [self makeUI];
        
    }
    return self;
}

-(void) makeUI {
    self.backgroundColor = [UIColor clearColor];
    self.layer.borderWidth = 1;
    self.layer.borderColor = [[UIColor groupTableViewBackgroundColor] CGColor];
    
    self.toolbar = [[UIToolbar alloc] initWithFrame:CGRectMake(0, 0, self.frame.size.width, 50)];
    self.toolbar.backgroundColor = [UIColor clearColor]; //[UIColor groupTableViewBackgroundColor];
    toolTitle = [[UILabel alloc] initWithFrame:CGRectMake(8, 0, self.toolbar.frame.size.width-50, self.toolbar.frame.size.height)];
    toolTitle.backgroundColor = [UIColor clearColor]; //[UIColor groupTableViewBackgroundColor];
    [self.toolbar addSubview:toolTitle];
    
    toolButton = [UIButton buttonWithType:UIButtonTypeSystem];
    toolButton.backgroundColor = [UIColor clearColor]; //[UIColor groupTableViewBackgroundColor];
    [toolButton setFrame:CGRectMake(self.toolbar.frame.size.width-50, 0, 50, self.toolbar.frame.size.height)];
    [toolButton setTitle:COMMON_DONE forState:UIControlStateNormal];
    [toolButton addTarget:self action:@selector(doneClick:) forControlEvents:UIControlEventTouchDown];
    [self.toolbar addSubview:toolButton];

    self.picker = [[UIPickerView alloc] initWithFrame:CGRectMake(0, 50, self.frame.size.width, 0)];
    self.picker.backgroundColor = [UIColor colorWithRed:0 green:0 blue:0 alpha:0.85];
    self.picker.autoresizingMask = UIViewAutoresizingFlexibleWidth;
    self.picker.showsSelectionIndicator = YES;
    self.picker.dataSource = self;
    self.picker.delegate = self;
    
    [self addSubview:self.toolbar];
    [self addSubview:self.picker];
}

-(void) setTitle:(NSString *)title {
    [toolTitle setText:title];
}

-(NSInteger) numberOfComponentsInPickerView:(UIPickerView *)pickerView {
    return 1;
}

-(NSInteger) pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
    return 65536;
}

-(CGFloat) pickerView:(UIPickerView *)pickerView rowHeightForComponent:(NSInteger)component {
    return pickerView.frame.size.height / 4;
}

-(CGFloat) pickerView:(UIPickerView *)pickerView widthForComponent:(NSInteger)component {
    return pickerView.frame.size.width;
}

-(UIView *) pickerView:(UIPickerView *)pickerView viewForRow:(NSInteger)row forComponent:(NSInteger)component reusingView:(UIView *)view {
    NSString * title = self.pickObjects[row % self.pickObjects.count];
    UILabel * lblView = (UILabel *)view;
    if (!lblView) {
        lblView = [[UILabel alloc] init];
        [lblView setBackgroundColor:[UIColor clearColor]];
        [lblView setTextColor:[UIColor whiteColor]];
        [lblView setTextAlignment:NSTextAlignmentCenter];
        [lblView setFont:[UIFont boldSystemFontOfSize:24]];
    }
    [lblView setText:title];
    return lblView;
}

-(void) doneClick: (id) sender {
    NSString * picked = self.pickObjects[[self.picker selectedRowInComponent:0] % self.pickObjects.count];
    if ([self.delegate respondsToSelector:@selector(pickDone:textField:)]) {
        [self.delegate pickDone:picked textField:self.txtResult];
    }
}

-(void) selectFirst {
    NSInteger selected = (65536 / 2) - (65536 % self.pickObjects.count) - 1;
    [self.picker selectRow:selected inComponent:0 animated:NO];
}

-(void) selectCurrent:(NSString *)text {
    NSInteger selected = (65536 / 2) - (65536 % self.pickObjects.count) - 1;
    for (NSInteger i=selected; i<selected+self.pickObjects.count; i++) {
        if ([self.pickObjects[i % self.pickObjects.count] isEqualToString:text]) {
            [self.picker selectRow:i inComponent:0 animated:NO];
            break;
        }
    }
}

@end

File: ios/YuGiOhCard/View/CardAttributePicker.swift
import UIKit

@objc protocol CardAttributePickerDelegate: NSObjectProtocol {
    @objc optional func pickDone(picked: String, textField: UITextField?)
}

class CardAttributePicker: UIView, UIPickerViewDataSource, UIPickerViewDelegate {

    var picker: UIPickerView?
    var toolbar: UIToolbar?
    var txtResult: UITextField?
    var pickObjects: NSArray?
    var delegate: CardAttributePickerDelegate?

    var toolTitle: UILabel?
    var toolButton: UIButton?
 
    override init(frame: CGRect) {
        super.init(frame: frame)
        self.makeUI()
  
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        self.makeUI()
    }
    
    func makeUI() {
        // self.backgroundColor = UIColor.blackColor()
        self.layer.borderWidth = 1
        self.layer.borderColor = UIColor.gray.cgColor
    
        self.toolbar = UIToolbar(frame: CGRect(x: 0, y: 0, width: self.frame.size.width, height: 50))
        self.toolbar!.tintColor = UIColor.white
        self.toolbar!.setBackgroundImage(UIImage(named: "popbg"), forToolbarPosition: UIBarPosition.any, barMetrics: UIBarMetrics.default)

        toolTitle = UILabel(frame: CGRect(x: 8, y: 0, width: self.toolbar!.frame.size.width-50, height: self.toolbar!.frame.size.height))
        toolTitle!.textColor = UIColor.white
        let splitLine = UIView(frame: CGRect(x: 0, y: self.toolbar!.frame.size.height, width: self.toolbar!.frame.size.width, height: 1))
        splitLine.backgroundColor = UIColor.gray
        
        self.toolbar!.addSubview(toolTitle!)
        self.toolbar!.addSubview(splitLine)
    
        toolButton = UIButton(type: UIButtonType.system)
        toolButton!.backgroundColor = UIColor.clear
        toolButton!.frame = (CGRect(x: self.toolbar!.frame.size.width-50, y: 0, width: 50, height: self.toolbar!.frame.size.height))
        toolButton!.setTitle(COMMON_DONE, for: [])
        toolButton!.addTarget(self, action:#selector(doneClick(sender:)), for:UIControlEvents.touchDown)
        self.toolbar!.addSubview(toolButton!)
    
        self.picker = UIPickerView(frame: CGRect(x: 0, y: 51, width: self.frame.size.width, height: self.frame.size.height - 50))
        self.picker!.backgroundColor = UIColor(red: 0, green: 0, blue: 0, alpha: 0.75)
        self.picker!.autoresizingMask = UIViewAutoresizing.flexibleWidth
        self.picker!.showsSelectionIndicator = true
        self.picker!.dataSource = self
        self.picker!.delegate = self
    
        self.addSubview(self.toolbar!)
        self.addSubview(self.picker!)
    }
    
    func setTitle(title: String) {
        toolTitle!.text = title
    }
    
    func numberOfComponents(in pickerView: UIPickerView) -> Int {
        return 1
    }
    
    func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        return 65535
    }
    
    func pickerView(_ pickerView: UIPickerView, rowHeightForComponent component: Int) -> CGFloat {
        return pickerView.frame.size.height / 4
    }
    
    func pickerView(_ pickerView: UIPickerView, widthForComponent component: Int) -> CGFloat {
        return pickerView.frame.size.width
    }
    
    func pickerView(_ pickerView: UIPickerView, viewForRow row: Int, forComponent component: Int, reusing view: UIView?) -> UIView {
        let title = self.pickObjects![row % self.pickObjects!.count] as! String
        var lblView: UILabel?
        if (view == nil) {
            lblView = UILabel()
            lblView!.backgroundColor = UIColor.clear
            lblView!.textColor = UIColor.white
            lblView!.textAlignment = NSTextAlignment.center
            lblView!.font = UIFont.boldSystemFont(ofSize: 24)
        } else {
            lblView = view as? UILabel
        }
        lblView!.text = title
        return lblView!
    }
    
    func doneClick(sender: AnyObject) {
        let picked = self.pickObjects![self.picker!.selectedRow(inComponent: 0) % self.pickObjects!.count] as! String
        self.delegate?.pickDone?(picked: picked, textField: self.txtResult)
    }
    
    func selectFirst() {
        let selected = (65536 / 2) - (65536 % self.pickObjects!.count) - 1
        self.picker!.selectRow(selected, inComponent: 0, animated: false)
    }
    
    func selectCurrent(text: String) {
        let selected = (65536 / 2) - (65536 % self.pickObjects!.count) - 1
        for i in selected ..< selected + self.pickObjects!.count {
            if ((self.pickObjects![i % self.pickObjects!.count] as! String) == text) {
                self.picker!.selectRow(i, inComponent:0, animated:false)
                break
            }
        }
    }

}
